main.go

```diff
@@ -201,49 +206,44 @@ func getActiveMiners(db *pg.DB, fp *filecoinPeer) (<-chan minerInfo, error) {
 		return nil, err
 	}
 
-	found := make(chan minerInfo)
-	var wg sync.WaitGroup
-	minersFound := 0
+	var (
+		found      = make([]minerInfo, len(activeMiners))
+		peerIDs    = make(chan minerWithDecodedPeerID, 50)
+		wg         = sync.WaitGroup{}
+		maxWorkers = runtime.GOMAXPROCS(0)
+		sem        = semaphore.NewWeighted(int64(maxWorkers))
+	)
+	for i := range activeMiners {
 
-	for i, activeMiner := range activeMiners {
-		i, activeMiner := i, activeMiner
-		wg.Add(1)
+		miner := activeMiners[i]
+
+		decodedPid, err := peer.Decode(miner.PeerID)
+		if err != nil {
+			log.Warnw("could not decode peer ID ", "ID", miner.PeerID, "error", err)
+		}
+		peerIDs <- minerWithDecodedPeerID{miner: miner, decodedPeerID: decodedPid}
 
+		wg.Add(1)
 		go func() {
 			defer wg.Done()
-
-			ctx, cancel := context.WithTimeout(context.Background(), time.Duration(15*time.Second))
-			defer cancel()
-
-			decodedPid, err := peer.Decode(activeMiner.PeerID)
-			if err != nil {
-				log.Warnw("could not decode peer ID ", "ID", activeMiner.PeerID, "error", err)
-				return
+			if err := sem.Acquire(fp.ctx, 1); err != nil {
+				log.Warnw("Failed to acquire semaphore: ", err)
```

Comment:
This is probably not a good pattern (I have not seen anything like this anywhere). It is overly complex in that it uses all a channel, a semaphore and a waitgroup, when a worker pattern can be solved with a channel.

It might work, in the sense that only 50 things are querying the DHT, but there will be also 50 items in the channel (I guess) and 50 goroutines waiting to access the semaphore (those that managed to put something in the channel), and another 50 executing (those that managed to Acquire the semaphore). Overly complex.